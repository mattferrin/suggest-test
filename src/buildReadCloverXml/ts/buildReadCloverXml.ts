import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
