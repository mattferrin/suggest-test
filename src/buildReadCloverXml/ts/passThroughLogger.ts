/* eslint-disable functional/no-conditional-statement */
/* eslint-disable functional/no-expression-statement */
/* eslint-disable functional/no-return-void */

import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export function passThroughLogger(
  summary: ReadCloverXmlOutput
): ReadCloverXmlOutput {
  switch (summary.tag) {
    case "errors":
      summary.errors.forEach((error) => {
        console.error(error);
      });
      break;
    case "summary":
      console.log(summary.summaryString);
      break;
  }

  return summary;
}
