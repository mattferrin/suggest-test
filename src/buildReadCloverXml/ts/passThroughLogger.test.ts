import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { passThroughLogger } from "./passThroughLogger";

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  summary: Parameters<typeof passThroughLogger>[0];
}): void {
  it(input.stabileSortId, () => {
    /** setup mocks */
    const logSpy = jest.spyOn(global.console, "log");
    const errorSpy = jest.spyOn(global.console, "error");

    /** unit under test */
    const result = passThroughLogger(input.summary);

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
        logCalls: logSpy.mock.calls,
        errorCalls: errorSpy.mock.calls,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "88d0259d-4eb5-4296-bd56-ccf7de479f59",
  comment: "displays errors in console AND passes summary result through",
  summary: {
    tag: "errors",
    errors: [new Error("expect this as console error")],
  },
});

snap({
  stabileSortId: "34693bef-26e9-403d-b4f6-038ebb5a02c4",
  comment: "displays summary in console AND passes summary result through",
  summary: {
    tag: "summary",
    summaryString: "expect this summary console log",
  } as ReadCloverXmlOutput,
});
