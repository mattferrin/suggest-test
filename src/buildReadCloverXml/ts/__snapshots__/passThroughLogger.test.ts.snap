// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`88d0259d-4eb5-4296-bd56-ccf7de479f59 1`] = `
{
  "displays errors in console AND passes summary result through": {
    "errorCalls": [
      [
        [Error: expect this as console error],
      ],
    ],
    "logCalls": [],
    "result": {
      "errors": [
        [Error: expect this as console error],
      ],
      "tag": "errors",
    },
  },
}
`;

exports[`34693bef-26e9-403d-b4f6-038ebb5a02c4 1`] = `
{
  "displays summary in console AND passes summary result through": {
    "errorCalls": [],
    "logCalls": [
      [
        "expect this summary console log",
      ],
    ],
    "result": {
      "summaryString": "expect this summary console log",
      "tag": "summary",
    },
  },
}
`;
